### PR TITLE
[quant][pt1] Add the dynamic quantized LSTM module

### PR DESCRIPTION
**Pull request by jianyuh**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25157 [quant][pt1] Add the dynamic quantized LSTM module**

Add the dynamic quantized LSTM module.

TODO (separate PRs):
- Serialization.
- Bias can be Null.

Differential Revision: [D16955662](https://our.internmc.facebook.com/intern/diff/D16955662/)

### CODE REVIEW

File: torch/nn/quantized/dynamic/modules/rnn.py [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+#         for weight in self._get_packed_weights():
+#             weight.set_(torch.zeros(torch.jit.annotate(List[int], []),
+#                         dtype=torch.uint8).detach())
+
+    @classmethod
+    def from_float(cls, mod):
+        # print(type(mod))
+        # assert LSTM or GRU
+        # assert type(mod) == torch.nn.quantized.dynamic.RNNBase.from_float,
+        #     'nn.quantized.dynamic.RNNBase.from_float only works for nn.LSTM'
```
Comment (jamesr66a) [inline torch/nn/quantized/dynamic/modules/rnn.py]: Why is this commented?

File: aten/src/ATen/native/RNN.cpp [hunk 2]
```diff
@@ -138,6 +166,52 @@ struct QuantizedCellParams {
   }
 };
 
+struct QuantizedCellParamsDynamic {
```
Comment (jamesr66a) [inline aten/src/ATen/native/RNN.cpp]: Is the purpose of this struct that we do not require explicit scale and zero point parameters? You should add a comment here explaining why we have both `QuantizedCellParams` and `QuantizedCellParamsDynamic`

**Review by jamesr66a — COMMENTED** [review]
This is a good start!

There's a ton of stuff commented, I'm assuming because some functionality doesn't work. We should add comments and/or issues documenting why these features are missing.

Additionally, I think there's some inconsistency in the code around whether LSTM and GRU or just LSTM is supported. We should clean up the code to ensure that everything is consistent.

File: aten/src/ATen/native/RNN.cpp [hunk 3]
```diff
@@ -3,10 +3,38 @@
...
+template <class... Args>
+inline std::vector<c10::IValue> callOp(
+    const char* func_name,
+    const char* overload_name,
+    Args... args) {
+  const c10::optional<c10::OperatorHandle> op_handle =
+      c10::Dispatcher::singleton().findSchema({func_name, overload_name});
+  assert(op_handle.has_value());
+  return callOp(op_handle.value(), args...);
+}
```
[the remark this answers is not in the document]
Comment (jianyuh) [reply aten/src/ATen/native/RNN.cpp]: @smessmer : Do you think if we can put this in C10 itself? Previously I chatted with you in the group, you pointed me to this solution in order to call c10 ops from ATen code (in C++). I took these code pieces from `kernel_functor_test.cpp`.

File: aten/src/ATen/native/RNN.cpp (hunk 2 above)
Comment (jianyuh) [reply aten/src/ATen/native/RNN.cpp]: That is right. Basically the Tensor in the arguments encapsulates the scale and the zero point in the struct in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/fbgemm_utils.h#L18-L23. I will add a comment here.

File: aten/src/ATen/native/RNN.cpp [hunk 4]
```diff
@@ -138,6 +166,52 @@ struct QuantizedCellParams {
   }
 };
 
+struct QuantizedCellParamsDynamic {
+  QuantizedCellParamsDynamic(
+      const Tensor& _w_ih, /* QTensor, qint8 */
+      const Tensor& _w_hh, /* QTensor, qint8 */
```
[the remark this answers is not in the document]
Comment (jianyuh) [reply aten/src/ATen/native/RNN.cpp]: Right. It is pre-packed. Will update the comment.

File: aten/src/ATen/native/RNN.cpp [hunk 5]
```diff
@@ -138,6 +166,52 @@ struct QuantizedCellParams {
...
+    TORCH_CHECK(false, "matmul is not supported with quantized cell params");
+  }
+  Tensor matmul_hh(Tensor h) const {
+    TORCH_CHECK(false, "matmul is not supported with quantized cell params");
+  }
+
+  template <class... Inputs>
+  inline std::vector<c10::IValue> makeStack(Inputs&&... inputs) {
+    return {std::forward<Inputs>(inputs)...};
+  }
```
[the remark this answers is not in the document]
Comment (jianyuh) [reply aten/src/ATen/native/RNN.cpp]: Will remove this.

File: aten/src/ATen/native/RNN.cpp [hunk 6]
```diff
@@ -138,6 +166,52 @@ struct QuantizedCellParams {
...
+
+  Tensor linear_ih(Tensor input_ih) const {
+    const auto kFuncName = "quantized::fbgemm_linear_dynamic";
+    const auto kOvrldName = "";
+    const std::vector<c10::IValue> output_ih_list = callOp(
+        kFuncName, kOvrldName, input_ih, w_ih, b_ih);
```
[the remark this answers is not in the document]
Comment (jianyuh) [reply aten/src/ATen/native/RNN.cpp]: Sure: will add an assert.

File: aten/src/ATen/native/RNN.cpp [hunk 7]
```diff
@@ -1014,6 +1101,56 @@ std::tuple<Tensor, Tensor, Tensor> quantized_lstm(
 
 }
 
+// Dynamic quantized implementations
+//
+// These implementations use FBGEMM to do the i2h and h2h linear layers with
+// an int8 quantized weight. This is advantageous in small-batch-size
+// scenarios where runtime is dominated by memory fetches of the weight matrix.
+
+std::tuple<Tensor, Tensor, Tensor> quantized_lstm_dynamic(
```
[the remark this answers is not in the document]
Comment (jianyuh) [reply aten/src/ATen/native/RNN.cpp]: Right: it is copied from there. Let me check if we can reuse one copy of the code

File: torch/nn/quantized/dynamic/modules/rnn.py [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+        return [getattr(self, name) for name in self._quantized_weights]
+
+#     # TODO: for some reason torch.jit.script_method causes a destruction of the
+#     # module to occur, which in turn frees the packed_ih object via its DataPtr
+#     # deleter. This is bizarre and should probably get fixed.
+#     # @torch._jit_internal.torch.jit.script_method
+#     # @torch.jit.script_method
+#     # _load_from_state_dict
+#     # __setstate__
+#     # load from the file
```
[the remark this answers is not in the document]
Comment (jianyuh) [reply torch/nn/quantized/dynamic/modules/rnn.py]: The serialization is not implemented in this PR. Maybe let me remove that in this PR and add a new PR to implement the serialization.

File: torch/nn/quantized/dynamic/modules/rnn.py (hunk 1 above)
Comment (jianyuh) [reply torch/nn/quantized/dynamic/modules/rnn.py]: Let me uncomment this.

File: torch/nn/quantized/dynamic/modules/rnn.py [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+        # else:
+        #     qRNNBase = GRU(mod.input_size, mod.hidden_size, mod.num_layers,
+        #                    mod.bias, mod.batch_first, mod.dropout, mod.bidirectional)
+
+        num_directions = 2 if mod.bidirectional else 1
+
+        assert mod.bias
+
+        # TODO: support more than just LSTM
+        if qRNNBase.mode != 'LSTM' and qRNNBase.mode != 'GRU':
```
[the remark this answers is not in the document]
Comment (jianyuh) [reply torch/nn/quantized/dynamic/modules/rnn.py]: I should only keep `if qRNNBase.mode != 'LSTM'` as GRU is not implemented yet.

**Review by jianyuh — COMMENTED** [review]
Thanks @jamesr66a ! Will work on the modification for this PR. As mentioned, I will separate the serialization as a new PR.

File: aten/src/ATen/native/RNN.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (jianyuh) [reply aten/src/ATen/native/RNN.cpp]: I will reuse the same `quantized_lstm` function with a new argument `use_dynamic` to check if we should use the new prepacked weight with the `PackedLinearWeight` struct.

File: aten/src/ATen/native/RNN.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jianyuh) [reply aten/src/ATen/native/RNN.cpp]: I now separate the c10 usage functions in an independent file: c10_utils.h. Hope this resolve the issue.

**Review by jamesr66a — APPROVED** [review]
LGTM!

File: aten/src/ATen/native/RNN.cpp [hunk 10]
```diff
@@ -138,6 +140,59 @@ struct QuantizedCellParams {
   }
 };
 
+// QuantizedCellParams vs. QuantizedCellParamsDynamic
+//
+// QuantizedCellParams uses the legacy
+// fbgemm_linear_int8_weighjt_fp32_activation API, which requires the explicit
```
[the remark this answers is not in the document]
Comment (jianyuh) [reply aten/src/ATen/native/RNN.cpp]: Fixed.

File: aten/src/ATen/native/RNN.cpp [hunk 11]
```diff
@@ -138,6 +140,59 @@ struct QuantizedCellParams {
...
+  Tensor matmul_hh(const Tensor& h) const {
+    TORCH_CHECK(false, "matmul is not supported with quantized cell params");
+  }
+
+  Tensor linear_ih(const Tensor& input_ih) const {
+    const auto kFuncName = "quantized::fbgemm_linear_dynamic";
+    const auto kOvrldName = "";
+    const std::vector<c10::IValue> output_ih_list =
+        callOp(kFuncName, kOvrldName, input_ih, w_ih, b_ih);
+    TORCH_CHECK(
```
[the remark this answers is not in the document]
Comment (jianyuh) [reply aten/src/ATen/native/RNN.cpp]: Fixed.

**Pull request closed without merging**